pkg/metrics/prometheus.go

```diff
@@ -0,0 +1,152 @@
+package metrics
+
+import (
+	"github.com/grafana/grafana/pkg/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"net/http"
+	"strings"
+)
+
+const (
+	namespace = "grafana"
+	subsystem = "backend"
+)
+
+type PrometheusPublisher struct {
+	CounterCollectors map[string]*prometheus.CounterVec
+	GaugeCollectors   map[string]*prometheus.GaugeVec
+}
+
+func CreatePrometheusPublisher() (*PrometheusPublisher, error) {
+	go func() {
+		http.Handle("/metrics", prometheus.Handler())
+		err := http.ListenAndServe(":9200", nil)
```

Comment:
> Dealing with raw counters and roll ups is very in accurate. Can not getting accurate sums over time. 

It's possible it was implemented incorrectly. Noone has done it yet for Prometheus data that I'm aware, but there's a bit more to time-based aggregation than dropping samples. Even then the magnitude of error of a naieve implementation shouldn't be too bad.

If you're using `rate` with Prometheus you will get as good an answer as we believe is possible, and what you'd be rolling up ultimately would be the output of that function which is basically a delta.

> Dropwizard??

It's a popular Java instrumentation library, also known as Codahale metrics in older incarnations. You appear to be using a fork of the Go port of it.

> Could you describe in what way they are not useful? 

"mean" is average latency since startup. Almost impossible to work with as it tends to average over too long a time. Can be useful as a shortcut on rare occasions.

"stddev" - I've never dug into exactly what this is, but stddev is not aggregatable.

"min"&"max" - Without knowing the time period involved these aren't very useful. Even then aggregation is tricky.

"p25"&friends - Here there's the question of time period, and non-aggregatability. They're also relatively inefficient (in Java they're about 10x slower than the Prometheus Histogram under contention).

You seem to be missing the other stats that are usually here: mean1, mean5 and mean15. These are exponentially decaying averages aiming for that many minutes. This means they are not particularly accurate, and they also turn out to be relatively expensive as they need to use a Mutex (in Java a Dropwizard Meter is about 8x slower than the Prometheus Counter and the Dropwizard Counter under contention).

> They are missing sum you say, sum of all the timing measurements? How would that be useful?

`rate(sum) / rate(count)` gives you the average latency over a given time period.

> How do prometheus exporter consume timings from other systems?

It depends on what the system offers, we have to make do with some less-than useful statistics sometimes or even do without. It's rare that a system has metrics with good semantics, thus it's much preferred to use our client libraries where practical.

> I would love to support prometheus out of the box but maybe better to have a node exporter plugin so it can scrape the metrics endpoint and process it however prometheus wants it :)

That we won't be doing :)